### PR TITLE
fix(40042): Convert to async function with decorator

### LINKED ISSUE #40042: Convert to async function with decorator
**TypeScript Version:**  3.9.7

**Search Terms:** async, codefix, decorator

**Code**

```ts
class Test {
  @Deco()
  test() {
    return fetch('www.typescript.com').then(x => x);
  }
}

function Deco() {
  return (_target, _key, descriptor: PropertyDescriptor) => descriptor;
}
```

**Expected behavior:**
After i run the "Convert to async function" codefix, the expcted result is:
```ts
  @Deco()
  async test() {
    const x = await fetch('www.typescript.com');
    return x;
  }
```
Or at least:
```ts
@Deco() async test() {
```

**Actual behavior:**
```ts
  async @Deco() test() {
    const x = await fetch('www.typescript.com');
    return x;
  }
```

**Playground Link:** [Playground Link](https://www.typescriptlang.org/play?#code/MYGwhgzhAEAqCmEAu0DeAoa0ACARewA9gBQCU0SiSZamW0ATvEgK4MB20AZs8ABbEA5AHdRAOiQBPAA6JgDAJbSkYogFtBpCX3jtiAD2gBeAHzR9pANx0AvujvouLdsCQLCnfERoYsTVhzQxAD6SGAMAObMADTQwQDW8JKxACZyisqEDABc0AAKDISyDFL4EPJKSFnkptBp5RlVDNY2QA)

**Related Issues:**

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #40042

### CODE REVIEW

File: src/services/codefixes/convertToAsyncFunction.ts [hunk 1]
```diff
@@ -73,7 +73,13 @@ namespace ts.codefix {
...
 
         // add the async keyword
-        changes.insertLastModifierBefore(sourceFile, SyntaxKind.AsyncKeyword, functionToConvert);
+        if (functionToConvert.modifiers) {
+            changes.insertModifierAt(sourceFile, functionToConvert.modifiers.end, SyntaxKind.AsyncKeyword, { prefix: " " });
+        }
+        else {
+            const pos = functionToConvert.decorators ? skipTrivia(sourceFile.text, functionToConvert.decorators.end) : functionToConvert.getStart(sourceFile);
+            changes.insertModifierAt(sourceFile, pos, SyntaxKind.AsyncKeyword, { suffix: " " });
+        }
```
Comment (sandersn) [inline src/services/codefixes/convertToAsyncFunction.ts:82]: I think it's a bit easier to read this way:

```suggestion
        const pos = 
            functionToConvert.modifiers ? functionToConvert.modifiers.end
            : functionToConvert.decorators ? skipTrivia(sourceFile.text, functionToConvert.decorators.end) 
            : functionToConvert.getStart(sourceFile);
        const options = functionToConvert.modifiers ? { prefix: " " } : { suffix: " " };
        changes.insertModifierAt(sourceFile, pos, SyntaxKind.AsyncKeyword, options);
```

Also writing it this way makes me wonder: does functionToConvert.modifiers.end also need to be wrapped in a skipTrivia call?

**Review by sandersn — CHANGES_REQUESTED** [review]
Possible refactor suggestion, plus a question whether skipTrivia is needed.

File: src/services/codefixes/convertToAsyncFunction.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/services/codefixes/convertToAsyncFunction.ts]: @sandersn Thanks for the review 👍. I think it depends. In the current implementation, the `async` keyword will be inserted right after the `public` modifier

```ts
class Foo {
    /* comment */ public async /* commnet */ method() {
        return fetch('a').then(x => x);
    }
}
````

If need to insert `async` after comment and before the function name, I think need to use `skipTrivia` 
```ts
class Foo {
    /* comment */ public /* commnet */ async method() {
        return fetch('a').then(x => x);
    }
}
````
Do I need to change it?

Comment (sandersn) [reply src/services/codefixes/convertToAsyncFunction.ts]: No, I think it's better the current way.

Comment (a-tarasyuk) [reply src/services/codefixes/convertToAsyncFunction.ts]: Ok

**Review by sandersn — APPROVED** [review]

**Comment by sandersn** [comment]
I'll merge this after we get the 4.1 RC branch sorted out.

**Pull request merged**